google/datalab/utils/commands/_commands.py

```diff
@@ -63,17 +76,88 @@ def create_args(line, namespace):
         args.append(arg)
     return args
 
-  def parse(self, line, namespace=None):
-    """Parses a line into a dictionary of arguments, expanding meta-variables from a namespace. """
+  def _get_args_to_parse_from_cell(self):
+
+    def _get_subparsers(p):
+      """Recursively get subparsers."""
+
+      subparsers = [
+        subparser
+        for action in p._actions
+        if isinstance(action, argparse._SubParsersAction)
+        for _, subparser in action.choices.items()
+      ]
+      ret = subparsers
+      for sp in subparsers:
+        ret += _get_subparsers(sp)
+      return ret
+
+    if not self._subparser_args_to_parse_from_cell:
+      subparsers = _get_subparsers(self)
+      subcommand_name_to_args = {}
+      for subparser in subparsers:
+        name = subparser.prog.split()[-1]
+        args_to_parse = subparser._args_to_parse_from_cell
+        subcommand_name_to_args[name] = args_to_parse
+      self._subparser_args_to_parse_from_cell = subcommand_name_to_args
+    return self._subparser_args_to_parse_from_cell
+
+  def parse(self, line, cell, namespace=None):
+    """Parses a line and cell into a dictionary of arguments, expanding variables from a namespace.
+
+    For each line parameters beginning with --, it also checks the cell content and see if it exists
+    there. For example, if "--config1" is a line parameter, it checks to see if cell dict contains
+    "config1" item, and if so, use the cell value. The "config1" item will also be removed from
+    cell content.
+
+    Args:
+      line: line content.
+      cell: cell content.
+      namespace: user namespace. If None, IPython's user namespace is used.
+
+    Returns:
+      A tuple of: 1. parsed config dict. 2. remaining cell after line parameters are extracted.
+    """
+
+    if namespace is None:
+      ipy = IPython.get_ipython()
+      namespace = ipy.user_ns
     try:
-      if namespace is None:
-        ipy = IPython.get_ipython()
-        namespace = ipy.user_ns
       args = CommandParser.create_args(line, namespace)
-      return self.parse_args(args)
+      subparser_name_to_args = self._get_args_to_parse_from_cell()
+      last_subcmd = next((x for x in reversed(args) if x in subparser_name_to_args), None)
```

Comment:
Because we don't want to get in the business of line args insertion. For simplicity, we always append at the end of the line.

None of our usage involves multiple subcommands in the same input.